### ISSUE TITLE
pixel for pixel images not matching on 100% hardness

### ISSUE THREAD

**Issue by Azalith** [arsenetar/dupeguru#1107]
**Describe the bug**
I have some sets of images that are identical pixel for pixel, checked using [Winmerge](https://winmerge.org/), but are not identical when compared in binary form, again using Winmerge. The images have sizes differing by less than a few hundred bytes.
These images do not match as duplicates using 100% hardness, scanning by content, in picture mode. I had expected them to match because they seem identical pixel for pixel (however, they match with 99% hardness).

I then tried setting up a separate test.
Image A and a copy of it, B, would be identified as duplicates (as expected) in picture mode, scanning contents, with hardness 100.
I then proceeded to open one of them in a text editor and added a space to the very end of the file. The two files then differed by a size of one byte, checked via windows file properties. Winmerge identified the images as still identical pixel for pixel, but different by that one character. Checking again in dupeguru with the same settings, they were not identified as duplicates with 100 hardness, but were able to be identified as duplicates with 99 hardness. Similar results.

**Additional context**
Some images I have are slight variations of each other, differing by very few pixels, and are marked as duplicates with 99% hardness as I would expect. The issue I am facing is that on 100% hardness, some sets of images that appear to be identical pixel for pixel, and are confirmed to be so when cross checked with winmerge, are not marked as duplicates. They would be if I used 99% hardness, but that would also include a lot of sets of images that are not pixel for pixel duplicates.

**Desktop (please complete the following information):**
 - OS: Windows 10 Pro
 - Dupeguru Version: 4.3.1

**Side note** : I may be naïve, and apologize for any inadequacies with this report. I have filled it out as sufficiently as I could reasonably think.

**Comment by Thibaud-P** [comment]
In matchblock.py:235, the percentage of similarity is forced back from 100 to 99 if the digest of the to pictures is different.
That seems to be your case, as at least file sizes differs.
You could try to build from sources deleting lines 235 and 236 in matchblock.py and check if it works as intended for you.
Perhaps this is a choice of the developers, and I understand it, as most users would put the slider to 100 if they want to find exact duplicates.
One solution would be to add an option "Only check pixels".

**Comment by arsenetar** [comment]
Yeah as @Thibaud-P this is by design to give an indication that "something is different" between the files.

**Comment by FredWahl** [comment]
I guess the best way to solve this is by introducing an api. Then you can write a dll with your own logic to determine if files are ”equal” in some sense. But then you need additional logic like only files that differ by x bytes.
